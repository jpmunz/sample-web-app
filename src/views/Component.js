import React from "react";
import { connect } from "react-redux";
import {
  Spinner,
  Button,
  InputGroup,
  FormControl,
  FormLabel
} from "react-bootstrap";
import {
  fetchFooServerValueRequest,
  getFooValue,
  getFooServerValue,
  setFooValue,
  getAsync
} from "src/state";
import autoBind from "react-autobind";
import "./Component.scss";

export class Component extends React.Component {
  constructor(props) {
    super(props);
    this.state = { fooValue: "" };
    autoBind(this);
  }

  componentDidMount() {
    this.props.fetchFooServerValueRequest({});
  }

  handleChangeFooValue(event) {
    this.setState({ fooValue: event.target.value });
  }

  handleClickSetFooValue() {
    this.props.setFooValue({ value: this.state.fooValue });
  }

  render() {
    if (this.props.request.inProgress) {
      return (
        <Spinner variant="primary" animation="border" role="status">
          <span className="sr-only">Loading...</span>
        </Spinner>
      );
    } else if (this.props.request.error) {
      return <div>{"Some error: " + this.props.request.error}</div>;
    } else {
      return (
        <div className="Component">
          <div>
            {"Got here from: " +
              (this.props.routeId ? "/foo/" + this.props.routeId : "/")}
          </div>
          <div>{"FooServerValue: " + this.props.fooServerValue}</div>
          <div>{"FooValue: " + this.props.fooValue}</div>
          <InputGroup>
            <FormLabel htmlFor="foo-value-input">Foo value</FormLabel>
            <FormControl
              id="foo-value-input"
              type="text"
              onChange={this.handleChangeFooValue}
              placeholder={this.props.fooValue}
              value={this.state.fooValue}
              aria-label="Foo value"
            />
            <InputGroup.Append>
              <Button variant="secondary" onClick={this.handleClickSetFooValue}>
                Update Foo Value
              </Button>
            </InputGroup.Append>
          </InputGroup>
        </div>
      );
    }
  }
}

export default connect(
  (state, props) => {
    return {
      fooValue: getFooValue(state),
      fooServerValue: getFooServerValue(state),
      request: getAsync(state, "foo")
    };
  },
  { setFooValue, fetchFooServerValueRequest }
)(Component);
