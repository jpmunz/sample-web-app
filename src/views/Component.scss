.Component {
  max-width: 500px;
}
