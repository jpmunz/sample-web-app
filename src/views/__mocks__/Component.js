import React from "react";

export const MockComponent = jest.fn();

MockComponent.mockReturnValue(<div>Mocked Component</div>);

const mock = jest.fn().mockImplementation(MockComponent);

export default mock;
