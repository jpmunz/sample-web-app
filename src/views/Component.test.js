import React from "react";
import { Component } from "./Component";
import { render, fireEvent } from "@testing-library/react";

const setFooValueMock = jest.fn();
const fetchFooServerValueRequestMock = jest.fn();

const renderComponent = function({
  fooValue = "foo value",
  routeId = null,
  fooServerValue = "foo server value",
  request = {}
}) {
  return render(
    <Component
      routeId={routeId}
      fooValue={fooValue}
      fooServerValue={fooServerValue}
      setFooValue={setFooValueMock}
      fetchFooServerValueRequest={fetchFooServerValueRequestMock}
      request={request}
    />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Request in progress", () =>
  renderTest({ request: { inProgress: true } }));
test("Request failed", () => renderTest({ request: { error: "failed" } }));
test("With a route id", () => renderTest({ routeId: "22" }));

test("Fetches are made on mount", () => {
  renderComponent({});
  expect(fetchFooServerValueRequestMock.mock.calls).toEqual([[{}]]);
});

test("Updating foo value", () => {
  const component = renderComponent({});

  fireEvent.change(component.getByLabelText("Foo value"), {
    target: { value: "bar" }
  });
  fireEvent.click(component.getByText("Update Foo Value"));

  expect(setFooValueMock.mock.calls).toEqual([[{ value: "bar" }]]);
});
