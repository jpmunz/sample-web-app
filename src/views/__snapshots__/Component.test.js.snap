// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request failed 1`] = `
<DocumentFragment>
  <div>
    Some error: failed
  </div>
</DocumentFragment>
`;

exports[`Request in progress 1`] = `
<DocumentFragment>
  <div
    class="spinner-border text-primary"
    role="status"
  >
    <span
      class="sr-only"
    >
      Loading...
    </span>
  </div>
</DocumentFragment>
`;

exports[`With a route id 1`] = `
<DocumentFragment>
  <div
    class="Component"
  >
    <div>
      Got here from: /foo/22
    </div>
    <div>
      FooServerValue: foo server value
    </div>
    <div>
      FooValue: foo value
    </div>
    <div
      class="input-group"
    >
      <label
        class="form-label"
        for="foo-value-input"
      >
        Foo value
      </label>
      <input
        aria-label="Foo value"
        class="form-control"
        id="foo-value-input"
        placeholder="foo value"
        type="text"
        value=""
      />
      <div
        class="input-group-append"
      >
        <button
          class="btn btn-secondary"
          type="button"
        >
          Update Foo Value
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
